app/src/main/java/app/revanced/integrations/shared/checks/Check.java

```diff
@@ -0,0 +1,128 @@
+                            str("revanced_check_environment_dialog_open_official_source_button"),
+                            (dialog, which) -> {
+                                final var intent = new Intent(Intent.ACTION_VIEW, GOOD_SOURCE);
+                                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+
+                                context.startActivity(intent);
+                            }
+                    ).setNegativeButton(
+                            str("revanced_check_environment_dialog_ignore_button"),
+                            (dialog, which) -> {
+                                final int current = Settings.CHECK_ENVIRONMENT_WARNING_ISSUED_COUNT.get();
+                                Settings.CHECK_ENVIRONMENT_WARNING_ISSUED_COUNT.save(current + 1);
+
+                                dialog.dismiss();
+                            }
+                    );
+
+    private final String failedReasonStringKey;
+
+    /**
+     * @param failedReasonStringKey The string key of the reason why the check failed to be shown to the user.
+     */
+    Check(String failedReasonStringKey) {
+        this.failedReasonStringKey = failedReasonStringKey;
+    }
+
+    protected abstract boolean run();
+
+    static boolean shouldRun() {
+        return Settings.CHECK_ENVIRONMENT_WARNING_ISSUED_COUNT.get() < 2;
+    }
+
+    static void disableForever() {
+        Logger.printDebug(() -> "Environment checks disabled forever.");
+
+        Settings.CHECK_ENVIRONMENT_WARNING_ISSUED_COUNT.save(Integer.MAX_VALUE);
+    }
+
+    @SuppressLint("NewApi")
+    static void issueWarning(Context context, Check... failedChecks) {
+        final var reasons = new StringBuilder();
+
+        reasons.append("<ul>");
+
+        for (var check : failedChecks) {
+            reasons.append("<li>").append(str(check.failedReasonStringKey));
+        }
+
+        reasons.append("</ul>");
+
+        final var finalMessage = Html.fromHtml(
+                String.format(str("revanced_check_environment_failed_message"), reasons),
+                FROM_HTML_MODE_COMPACT
+        );
+        AlertDialog dialog = CHECK_FAILED_DIALOG_BUILDER.apply(context)
+                .setMessage(finalMessage).create();
+
+        var dismissButton = dialog.getButton(DialogInterface.BUTTON_NEGATIVE);
+        // TODO: setEnabled is called on null for some reason.
+        // dismissButton.setEnabled(false);
+
+        dialog.show();
+
+        ((TextView)dialog.findViewById(android.R.id.message)).setMovementMethod(LinkMovementMethod.getInstance());
+
+        // TODO: This does not work for some reason.
+        // Use a delay to allow the activity to finish initializing.
+        // Otherwise, if device is in dark mode the dialog is shown with wrong color scheme.
+        Utils.runOnMainThreadDelayed(getCountdownRunnable(dismissButton), 100);
+    }
+
+    private static Runnable getCountdownRunnable(Button dismissButton) {
```

Comment:
For 2 seconds the button text does not change for that reason to not distract the user starting to read. Keeping the button static could result in people just clicking on the other button in despair to close the dialog and not understand why they are on a different site now and what the dialog was about.

Reply:
I am certain this will make people screenshot the dialog and ask in support why they can't exit the screen 😆 (probably unlikely) but i guess it's fine since most people will simply checkout the dialog content.